### PR TITLE
Transpose sinking on passes for Concat, Split and binary eltwise operations

### PR DESCRIPTION
**Pull request by evkotov**
### Details:
 - transpose sinking group of transformations
 - binary eltwise operations
 - concat
 - split
 - use ngraph MatcherPass
 - forward and backward propagation
 - unit tests

### Tickets:
 - 92221
 - 94139

### CODE REVIEW

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+    auto binary_label_left = ov::pass::pattern::wrap_type<ov::op::util::BinaryElementwiseArithmetic>({transpose_label,
+                                                                                                      ov::pass::pattern::any_input()});
+
+    auto binary_label_right = ov::pass::pattern::wrap_type<ov::op::util::BinaryElementwiseArithmetic>({ov::pass::pattern::any_input(),
+                                                                                                 transpose_label});
+    auto binary_label = std::make_shared<ngraph::pattern::op::Or>(ngraph::OutputVector{binary_label_left, binary_label_right});
+
+    ov::matcher_pass_callback matcher_pass_callback = [=](ov::pass::pattern::Matcher& m) {
+        auto binary = m.get_match_root();
+        auto transpose = GetFirstTransposeInput(binary.get());
```
[the remark this answers is not in the document]
Comment (evkotov) [reply src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:346]: No it doesn't work. If there are two transpose inputs it founds not the first transpose input.

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+};
+
+bool IfConcatHasTransposeInputs(const ov::Output<ov::Node>& output) {
+    return FindFirstInputIf(output.get_node(), IfTranpose()) >= 0;
+}
+
+NodePtr GetFirstTransposeInput(const ov::Node * node) {
+    const int index = FindFirstInputIf(node, IfTranpose());
+    if (index < 0)
+        return {};
```
[the remark this answers is not in the document]
Comment (evkotov) [reply src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:60]: No. This function is also used in IfNodeHasTransposeInputs to check if Concat has any input with transposes. If no we just check that return is empty shared_ptr<Node>. Throwing exception and than catching it later is not the best idea.

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+    }
+
+    return -1;
+}
+
+class IfTranpose {
+public:
+    IfTranpose() = default;
+    bool operator()(const ov::Node * node) const {
+        return dynamic_cast<const ov::opset9::Transpose*>(node) != nullptr;
```
[the remark this answers is not in the document]
Comment (evkotov) [reply src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:49]: I wanted to separate traversing inputs and predicate that input is transpose.

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+    return 0;
+}
+
+ov::AxisVector GetTransposeOrder(NodePtr transpose)
+{
+    auto transpose_constant_node = ov::as_type_ptr<ov::opset9::Constant>(transpose->input_value(1).get_node_shared_ptr());
+    return transpose_constant_node->get_axis_vector_val();
+}
+
+ov::AxisVector ReverseTransposeOrder(const ov::AxisVector & axis_order)
```
[the remark this answers is not in the document]
Comment (evkotov) [reply src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:298]: I will check it. But on the first glance std::reverse is just reversing vector elements in backward order. But here we need to reverse a transpose permutation. I think it is not the same

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+    for (auto & node : nodes)
+        for (auto & output : node->outputs())
+            outputs.push_back(output);
+
+    return outputs;
+}
+
+// --------------------------------------------------------------------------------------
+
+struct TrasposeInputInfo {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:34]:
```suggestion
struct TransposeInputInfo {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+    std::shared_ptr<ov::opset9::Transpose> transpose;
+    std::shared_ptr<ov::opset9::Constant> transpose_const;
+    int input_idx;
+};
+
+TrasposeInputInfo GetFirstTransposeInput(NodePtr node) {
+    for (int input_idx = 0; input_idx < node->get_input_size(); ++input_idx) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:41]:
```suggestion
    for (size_t input_idx = 0; input_idx < node->get_input_size(); ++input_idx) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+        {
+            TrasposeInputInfo input_info;
+            input_info.transpose = transpose_node;
+            input_info.transpose_const = constant_node;
+            input_info.input_idx = input_idx;
+            return input_info;
+        }
+    }
+
+    throw std::runtime_error("no transpose input found");
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:58]: Do we need it?

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+
+    return TrasposeInputInfo();
+}
+
+bool IfNodeHasTransposeInputs(const ov::Output<ov::Node>& output) {
+    for (int input_idx = 0; input_idx < output.get_node()->get_input_size(); ++input_idx) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:64]:
```suggestion
    for (size_t input_idx = 0; input_idx < output.get_node()->get_input_size(); ++input_idx) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+        return true;
+    }
+
+    return false;
+}
+
+// --------------------------------------------------------------------------------------
+
+ov::AxisVector ReverseTransposeOrder(const ov::AxisVector & axis_order)
+{
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:81]: style {

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 10]
```diff
@@ -0,0 +1,525 @@
...
+    ov::AxisVector out(axis_order.size());
+    for (size_t i = 0; i < axis_order.size(); i++) {
+        out.at(axis_order[i]) = i;
+    }
+    return out;
+}
+
+}
+
+void SwapOutputNames(ov::Output<ov::Node> output1, ov::Output<ov::Node> output2) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:91]:
```suggestion
void SwapOutputNames(ov::Output<ov::Node>& output1, ov::Output<ov::Node>& output2) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 11]
```diff
@@ -0,0 +1,525 @@
...
+
+    auto m = std::make_shared<ov::pass::pattern::Matcher>(split_label, matcher_name);
+    register_matcher(m, matcher_pass_callback);
+}
+
+struct OutputTranspose {
+    ov::opset9::Transpose * transpose;
+    ov::opset9::Constant * transpose_const;
+    int output_idx;
+    int input_idx;
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:414]:
```suggestion
    size_t input_idx;
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 12]
```diff
@@ -0,0 +1,525 @@
...
+    };
+
+    auto m = std::make_shared<ov::pass::pattern::Matcher>(split_label, matcher_name);
+    register_matcher(m, matcher_pass_callback);
+}
+
+struct OutputTranspose {
+    ov::opset9::Transpose * transpose;
+    ov::opset9::Constant * transpose_const;
+    int output_idx;
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:413]:
```suggestion
    size_t output_idx;
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 13]
```diff
@@ -0,0 +1,525 @@
...
+        const size_t tranpose_input_index = transpose_input_info.input_idx;
+        const ov::element::Type transpose_element_type = transpose_input_info.transpose_const->get_element_type();
+
+        for (size_t i = 0; i < binary->get_input_size(); ++i) {
+            auto input_node = binary->input_value(i);
+            if (i == tranpose_input_index) {
+                auto transpose_parent = input_node.get_node()->input_value(0);
+                binary->input(i).replace_source_output(transpose_parent);
+            }
+            else {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:137]: style 
```
} else {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 14]
```diff
@@ -0,0 +1,525 @@
...
+
+} // namespace
+
+ngraph::pass::TransposeSinkingConcatForward::TransposeSinkingConcatForward() {
+    MATCHER_SCOPE(TransposeSinkingConcatForward);
+
+    auto concat_label = ov::pass::pattern::wrap_type<ov::opset9::Concat>(IfNodeHasTransposeInputs);
+
+    ov::matcher_pass_callback matcher_pass_callback = [=](ov::pass::pattern::Matcher& m) {
+        const auto & pattern_to_output = m.get_pattern_value_map();
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:240]: Let's use the same & aligment in code

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 15]
```diff
@@ -0,0 +1,525 @@
...
+
+        // insert output transposes
+        for (size_t i = 0; i < split->get_output_size(); ++i) {
+            auto new_transpose_const = std::make_shared<ov::opset9::Constant>(transpose_element_type,
+                                                                              ov::Shape{transpose_axis_order.size()},
+                                                                              transpose_axis_order);
+            auto new_transpose = std::make_shared<ov::opset9::Transpose>(split, new_transpose_const);
+
+            auto split_consumers = split->output(i).get_target_inputs();
+            for (auto consumer: split_consumers) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:389]:
```suggestion
            for (auto& consumer: split_consumers) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 16]
```diff
@@ -0,0 +1,525 @@
...
+                                                                              ov::Shape{reversed_traspose_axis_order.size()},
+                                                                              reversed_traspose_axis_order);
+            auto new_transpose = std::make_shared<ov::opset9::Transpose>(split, new_transpose_const);
+
+            for (auto input : split->get_output_target_inputs(output_idx)) {
+                if (output_idx != output_transpose.output_idx || input.get_index() != output_transpose.input_idx) {
+                    input.replace_source_output(new_transpose);
+                } else {
+                    auto transpose_consumers = input.get_node()->output(0).get_target_inputs();
+                    for (auto consumer: transpose_consumers) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:514]:
```suggestion
                    for (auto& consumer: transpose_consumers) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 17]
```diff
@@ -0,0 +1,525 @@
...
+        }
+
+        auto binary_consumers = binary->output(0).get_target_inputs();
+
+        auto new_transpose_const = std::make_shared<ov::opset9::Constant>(transpose_element_type,
+                                                                          ov::Shape{transpose_axis_order.size()},
+                                                                          transpose_axis_order);
+        auto new_transpose = std::make_shared<ov::opset9::Transpose>(binary, new_transpose_const);
+
+        for (auto consumer: binary_consumers) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:156]:
```suggestion
        for (auto& consumer: binary_consumers) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 18]
```diff
@@ -0,0 +1,525 @@
...
+
+            binary->input(i).replace_source_output(new_transpose->output(0));
+
+            ov::copy_runtime_info(input_node.get_node_shared_ptr(), {new_transpose, new_transpose_const});
+
+            register_new_node(new_transpose);
+        }
+
+        auto transpose_consumers = transpose->output(0).get_target_inputs();
+        for (auto consumer: transpose_consumers) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:209]:
```suggestion
        for (auto& consumer: transpose_consumers) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 19]
```diff
@@ -0,0 +1,525 @@
...
+        }
+
+        auto binary_consumers = concat->output(0).get_target_inputs();
+
+        auto new_transpose_const = std::make_shared<ov::opset9::Constant>(transpose_element_type,
+                                                                          ov::Shape{transpose_axis_order.size()},
+                                                                          transpose_axis_order);
+        auto new_transpose = std::make_shared<ov::opset9::Transpose>(concat, new_transpose_const);
+
+        for (auto consumer: binary_consumers) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:277]:
```suggestion
        for (auto& consumer: binary_consumers) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 20]
```diff
@@ -0,0 +1,525 @@
...
+
+            concat->input(i).replace_source_output(new_transpose->output(0));
+
+            ov::copy_runtime_info(input_node.get_node_shared_ptr(), {new_transpose, new_transpose_const});
+
+            register_new_node(new_transpose);
+        }
+
+        auto transpose_consumers = transpose->output(0).get_target_inputs();
+        for (auto consumer: transpose_consumers) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:333]:
```suggestion
        for (auto& consumer: transpose_consumers) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 21]
```diff
@@ -0,0 +1,525 @@
...
+
+struct OutputTranspose {
+    ov::opset9::Transpose * transpose;
+    ov::opset9::Constant * transpose_const;
+    int output_idx;
+    int input_idx;
+};
+
+OutputTranspose GetOutputTransposes(NodePtr node) {
+    for (int output_idx = 0; output_idx < node->get_output_size(); ++output_idx) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:418]:
```suggestion
    for (size_t output_idx = 0; output_idx < node->get_output_size(); ++output_idx) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 22]
```diff
@@ -0,0 +1,525 @@
...
+        for (auto input : node->get_output_target_inputs(output_idx)) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:419]:
```suggestion
        for (auto& input : node->get_output_target_inputs(output_idx)) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 23]
```diff
@@ -0,0 +1,525 @@
...
+
+    throw std::runtime_error("cannot find output transpose");
+
+    return OutputTranspose();
+}
+
+bool HasOutputTranposes(const ov::Output<ov::Node>& output) {
+    NodePtr node = output.get_node_shared_ptr();
+
+    for (int output_idx = 0; output_idx < node->get_output_size(); ++output_idx) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:446]:
```suggestion
    for (size_t output_idx = 0; output_idx < node->get_output_size(); ++output_idx) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 24]
```diff
@@ -0,0 +1,525 @@
...
+        }
+
+        // update split axis
+        auto new_split_axis_const = std::make_shared<ov::opset9::Constant>(split_axis_constant->get_element_type(),
+                                                                           ov::Shape{},
+                                                                           reversed_transposed_split_axis);
+        split->input(1).replace_source_output(new_split_axis_const);
+
+        // update split output
+        for (int output_idx = 0; output_idx < split->get_output_size(); ++output_idx) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:503]:
```suggestion
        for (size_t output_idx = 0; output_idx < split->get_output_size(); ++output_idx) {
```

File: src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp [hunk 25]
```diff
@@ -0,0 +1,525 @@
...
+            auto new_transpose_const = std::make_shared<ov::opset9::Constant>(transpose_element_type,
+                                                                              ov::Shape{reversed_traspose_axis_order.size()},
+                                                                              reversed_traspose_axis_order);
+            auto new_transpose = std::make_shared<ov::opset9::Transpose>(split, new_transpose_const);
+
+            for (auto input : split->get_output_target_inputs(output_idx)) {
```
Comment (nkogteva) [inline src/common/transformations/src/transformations/common_optimizations/transpose_sinking_binary.cpp:509]:
```suggestion
            for (auto& input : split->get_output_target_inputs(output_idx)) {
```

**Review by nkogteva — COMMENTED** [review]
In general, LGTM

File: src/common/transformations/include/transformations/common_optimizations/transpose_sinking_binary.hpp [hunk 26]
```diff
@@ -0,0 +1,58 @@
...
+    TransposeSinkingConcatForward();
+};
+
+class ngraph::pass::TransposeSinkingConcatBackward : public ov::pass::MatcherPass {
+public:
+    OPENVINO_RTTI("ngraph::pass::TransposeSinkingConcatBackward", "0");
+    TransposeSinkingConcatBackward();
+};
+
+class ngraph::pass::TransposeSinkingSplitForward : public ov::pass::MatcherPass {
```
Comment (itikhono) [inline src/common/transformations/include/transformations/common_optimizations/transpose_sinking_binary.hpp:48]: Can we combine TransposeSinkingSplit and TransposeSInkingBinary?
Does Split has a specific except number of outputs?

**Review by itikhono — CHANGES_REQUESTED** [review]

**Comment by itikhono** [comment]
PR has been re-opened https://github.com/openvinotoolkit/openvino/pull/13718

**Pull request closed without merging**